### PR TITLE
[FLINK-8845][state] Introduce RocksDBWriteBatchWrapper to improve performance for recovery in RocksDB backend

### PR DESCRIPTION
**Pull request by sihuazhou**
## What is the purpose of the change

This PR addresses [FLINK-8845](https://issues.apache.org/jira/browse/FLINK-8845), which attempts to use `WriteBatch` to improve the performance for loading data into RocksDB. It's inspired by [RocksDB FAQ](https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ).

## Brief change log

  - *Introduce `RocksDBWriteBatchWrapper` to load data into RocksDB in bulk*

## Verifying this change

- Introduce `RocksDBWriteBatchWrapperTest.java` to guard `RocksDBWriteBatchWrapper`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation
none

### CODE REVIEW

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	private final int capacity;
+
+	private int currentSize;
+
+	public RocksDBWriteBatchWrapper(@Nonnull RocksDB rocksDB,
+									@Nonnull WriteOptions options,
+									int capacity) {
+
+		Preconditions.checkArgument(capacity >= MIN_CAPACITY && capacity <= MAX_CAPACITY,
+			"capacity should at least greater than 100");
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java]: how is the capacity range determined - is it recommended by RocksDB?

the msg should be: "capacity should be between " + MIN + " and " + MAX

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+		this.db = rocksDB;
+		this.options = options;
+		this.capacity = capacity;
+		this.batch = new WriteBatch(this.capacity);
+		this.currentSize = 0;
+	}
+
+	public void put(ColumnFamilyHandle handle, byte[] key, byte[] value) throws RocksDBException {
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java]: need synchronization on put() and flush()

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+	public void flush() throws RocksDBException {
+		this.db.write(options, batch);
+		batch.clear();
+		currentSize = 0;
+	}
+
+	@Override
+	public void close() throws RocksDBException {
+		if (batch != null) {
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java]: can batch be null?

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+			// valid result
+			for (Tuple2<byte[], byte[]> item : data) {
+				Assert.assertArrayEquals(item.f1, db.get(handle, item.f0));
+			}
+		}
+	}
+
+	@Test
+	@Ignore
+	public void benchMark() throws Exception {
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: Need to move the benchmark test to `org.apache.flink.contrib.streaming.state.benchmark` package.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		benchMarkHelper(10_000, false, WRITETYPE.PUT);
+
+		benchMarkHelper(100_000, true, WRITETYPE.PUT);
+		benchMarkHelper(100_000, false, WRITETYPE.PUT);
+
+		benchMarkHelper(1_000_000, true, WRITETYPE.PUT);
+		benchMarkHelper(1_000_000, false, WRITETYPE.PUT);
+
+		// put with disableWAL=true VS write batch with disableWAL=false
+		System.out.println("--> put with disableWAL=true VS write batch with disableWAL=false <--");
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: replace console output with logging, you can refer to `RocksDBListStatePerformanceTest.java`

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+			case PUT:
+				try (RocksDB db = RocksDB.open(folder.newFolder().getAbsolutePath());
+					WriteOptions options = new WriteOptions().setDisableWAL(disableWAL);
+					ColumnFamilyHandle handle = db.createColumnFamily(new ColumnFamilyDescriptor("test".getBytes()))) {
+					long t1 = System.currentTimeMillis();
+					for (Tuple2<byte[], byte[]> item : data) {
+						db.put(handle, options, item.f0, item.f1);
+					}
+					long t2 = System.currentTimeMillis();
+					System.out.println(" put cost:" + (t2 - t1) + " ms");
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: ditto

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+					WriteOptions options = new WriteOptions().setDisableWAL(disableWAL);
+					ColumnFamilyHandle handle = db.createColumnFamily(new ColumnFamilyDescriptor("test".getBytes()));
+					RocksDBWriteBatchWrapper writeBatchWrapper = new RocksDBWriteBatchWrapper(db, options, 200)) {
+					long t1 = System.currentTimeMillis();
+					for (Tuple2<byte[], byte[]> item : data) {
+						writeBatchWrapper.put(handle, item.f0, item.f1);
+					}
+					writeBatchWrapper.flush();
+					long t2 = System.currentTimeMillis();
+					System.out.println(" write batch cost:" + (t2 - t1) + " ms");
```
Comment (bowenli86) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: ditto

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java (hunk 4 above)
Comment (sihuazhou) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: 👍

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java (hunk 5 above)
Comment (sihuazhou) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: 👍

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java (hunk 6 above)
Comment (sihuazhou) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: 👍

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java (hunk 7 above)
Comment (sihuazhou) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapperTest.java]: 👍

**Comment by sihuazhou** [comment]
@bowenli86 Thanks a lot for you reviews, I've addressed all your comments, I'm waiting the 1.5 to be released, after that maybe @StefanRRichter could also have a look at this.

**Comment by sihuazhou** [comment]
A micro-benchmark for this: @StefanRRichter @StephanEwen 
```
-----------------------> With disableWAL is false <-----------------
Number of values added | time for Put		|  time for WriteBach | performance improvement of WriteBatch over Put
1000			10146397 ns		3546287 ns			2.86x
10000			118227077 ns		26040222 ns			4.54x
100000			1838593196 ns		375053755 ns			4.9x
1000000			8844612079 ns		2014077396 ns			4.39x

-----------------------> With disableWAL is true <-----------------
1000			3955204 ns		2429725 ns			1.62x
10000			25618237 ns		16440113 ns			1.55x
100000			289153346 ns		183712685 ns			1.57x
1000000			2886298967 ns		1768688571 ns			1.63x
```

**Comment by StefanRRichter** [comment]
@sihuazhou thanks for this nice contribution. LGTM 👍 Will merge.

**Pull request closed without merging**